backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/InventoryLineModularContractHandler.java

```diff
@@ -0,0 +1,191 @@
+			@NonNull final FlatrateTermId modularContractId)
+	{
+		final I_M_Inventory inventory = inventoryBL.getById(InventoryId.ofRepoId(inventoryLine.getM_Inventory_ID()));
+
+		final InventoryId reversalId = InventoryId.ofRepoIdOrNull(inventory.getReversal_ID());
+		if (reversalId != null)
+		{
+			return Optional.empty();
+		}
+
+		final ModularContractSettings modularContractSettings = modularContractSettingsDAO.getByFlatrateTermIdOrNull(modularContractId);
+		if (modularContractSettings == null)
+		{
+			throw new AdempiereException("Invalid contract missing settings")
+					.appendParametersToMessage()
+					.setParameter("M_InventoryLine_ID", inventoryLine.getM_InventoryLine_ID())
+					.setParameter("ModularContractID", modularContractId);
+		}
+
+		final Optional<ModularContractTypeId> modularContractTypeId = modularContractSettings.getModuleConfigs()
+				.stream()
+				.filter(config -> config.isMatchingClassName(InventoryLineModularContractHandler.class.getName()))
+				.map(ModuleConfig::getModularContractType)
+				.map(ModularContractType::getId)
+				.findFirst();
+
+		final Quantity quantity = inventoryBL.getMovementQty(inventoryLine);
+
+		final String description = msgBL.getMsg(MSG_DESCRIPTION, ImmutableList.of(quantity.abs().toString()));
+
+		final I_C_Flatrate_Term modularContractRecord = flatrateDAO.getById(modularContractId);
+
+		return modularContractTypeId.map(contractTypeId -> LogEntryCreateRequest.builder()
+				.contractId(modularContractId)
+				.productId(ProductId.ofRepoId(inventoryLine.getM_Product_ID()))
+				.referencedRecord(TableRecordReference.of(I_M_InventoryLine.Table_Name, inventoryLine.getM_InventoryLine_ID()))
+				.invoicingBPartnerId(BPartnerId.ofRepoIdOrNull(modularContractRecord.getBill_BPartner_ID()))
+				.warehouseId(WarehouseId.ofRepoIdOrNull(inventory.getM_Warehouse_ID()))
+				.documentType(LogEntryDocumentType.INVENTORY)
+				.soTrx(SOTrx.PURCHASE)
+				.quantity(quantity)
+				.transactionDate(LocalDateAndOrgId.ofTimestamp(inventory.getMovementDate(),
+															   OrgId.ofRepoId(inventoryLine.getAD_Org_ID()),
+															   orgDAO::getTimeZone))
+				.year(modularContractSettings.getYearAndCalendarId().yearId())
+				.description(description)
+				.modularContractTypeId(contractTypeId)
+				.build());
+	}
+
+	@Override
+	public @NonNull Optional<LogEntryReverseRequest> createLogEntryReverseRequest(
+			@NonNull final I_M_InventoryLine inventoryLine,
+			@NonNull final FlatrateTermId modularContractId)
+	{
+		final I_M_Inventory inventory = inventoryBL.getById(InventoryId.ofRepoId(inventoryLine.getM_Inventory_ID()));
+
+		final InventoryId reversalId = InventoryId.ofRepoIdOrNull(inventory.getReversal_ID());
+		if (reversalId != null)
+		{
+			return Optional.empty();
+		}
+
+		return Optional.of(LogEntryReverseRequest.builder()
+								   .referencedModel(TableRecordReference.of(inventoryLine))
+								   .flatrateTermId(modularContractId)
+								   .build());
```

Comment:
💡 can be done later

I think we should provide a different description for the reversal log line, to make it obvious this inventory was reversed.

Reply:
The used description is `A shortage/excess stock of {0} was counted during inventory.`, and the parameter is always positive (so that it is applicable in both cases - complete and reverse). Shall we change the description and use a different one?

Reply:
I see your point regarding the description. I am not sure how detailed our customer would like that description to be but I have the feeling they would like to see if the document was reversed. I can bring it up in a future meeting with Mark and will document my findings in another issue.
For now, let's just leave it like it is